### PR TITLE
SelectControl: Infer `value` type from `options`

### PR DESCRIPTION
**Pull request by mirka**
Updated take on #60293 (props @mikeybinns)

## What?

Tighten the TypeScript type checking on `SelectControl`, so the `value` type matches in the `value`, `options`, and `onChange` props.

See Dev Notes at the bottom for how this looks like to a consumer.

## Why?

It can catch potential bugs.

## How?

Adding a generic that can either be inferred from the `options` array (if it is immutable), or be set explicitly.

## Testing Instructions

See some of the static type tests in the unit test file, and play around by adding your own cases and see what errors.

---

### Stricter type checking in `SelectControl`

In TypeScript codebases, `SelectControl` will now be able to type check whether the `value` type matches in the `value`, `options`, and `onChange` props.

<details>
<summary>This will be inferred automatically if the <code>options</code> array is declared as immutable.</summary>
<p>

```tsx
<SelectControl
	// @ts-expect-error "baz" is not in `options` array
	value="baz"
	options={ [ // array literal that cannot be mutated
		{
			value: 'foo',
			label: 'Foo',
		},
		{
			value: 'bar',
			label: 'Bar',
		},
	] }
/>;
```

```tsx
const options = [
	{
		value: 'foo',
		label: 'Foo',
	},
	{
		value: 'bar',
		label: 'Bar',
	},
] as const; // const assertion to mark as readonly

<SelectControl
	// @ts-expect-error "baz" is not in `options` array
	value="baz"
	options={ options }
/>;
```

</p>
</details>

<details>
<summary>If the <code>options</code> array is mutable, <code>value</code> will remain a simple <code>string</code> type.</summary>
<p>

This is effectively the same type checking behavior as before.

```tsx
const options = [
	{
		value: 'foo',
		label: 'Foo',
	},
	{
		value: 'bar',
		label: 'Bar',
	},
];

<SelectControl
	// No error, because `value` has type `string`
	value="baz"
	options={ options }
/>;
```

</p>
</details>

<details>
<summary>You can also set an explicit <code>value</code> type.</summary>
<p>

```tsx
<SelectControl< 'foo' | 'bar' >
	// @ts-expect-error "baz" is not in `'foo' | 'bar'`
	value="baz"
	// @ts-expect-error "baz" is not in `'foo' | 'bar'`
	options={ [ { value: 'baz', label: 'Baz' } ] }
/>;
```

</p>
</details>

### CODE REVIEW

File: packages/components/src/date-time/time/index.tsx
Comment (mirka) [inline packages/components/src/date-time/time/index.tsx:1]: The changes we have to make in this file are what make me a bit hesitant about adding this `value` type inference.  The stricter types do not break runtime code obviously, but we will be forcing some consumers to deal with this new strictness, which may seem like overkill depending on who you ask.

I _think_ we can get away with it, because the strictness will only kick in if the `options` array is defined as immutable (either a literal array in the component call, or with an `as const`). For example, there is a [potential bug](https://github.com/WordPress/gutenberg/blob/f50e045c1ce75acd026ab760891cd5bf57234fd3/packages/components/src/custom-gradient-picker/index.tsx#L112-L114) in `CustomGradientPicker` but it isn't caught with the new strictness because `GRADIENT_OPTIONS` is mutable.

We don't have a ton of data points in the type checked realm of this repo, but it seems like `TimePicker` is the only "false positive" in the sense that it didn't surface a potential bug. (I just want to be sure that this type tightening is a net positive and not a nuisance 😅)

File: packages/components/src/query-controls/index.tsx
Comment (mirka) [inline packages/components/src/query-controls/index.tsx:92]: This was an actual bug.

File: packages/components/src/select-control/index.tsx
Comment (mirka) [inline packages/components/src/select-control/index.tsx]: These type assertions should be safe, as they will always match the `options` types.

File: packages/components/src/select-control/types.ts
Comment (mirka) [inline packages/components/src/select-control/types.ts]: Adding a `readonly` here so it allows immutable arrays. (Otherwise it would be a type error when passing an immutable array.)

File: packages/components/src/select-control/types.ts
Comment (mirka) [inline packages/components/src/select-control/types.ts]: `NoInfer<>` is a [new utility type](https://devblogs.microsoft.com/typescript/announcing-typescript-5-4/#the-noinfer-utility-type) added in TS 5.4, which tells TS not to use this field to infer the generic type. We need this because we want it to infer from the values in the `options` array, not the `value` prop.

This kind of "don't infer" signaling [used to be done](https://www.typescriptlang.org/play/?ts=5.4.5#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQdMAnmFnAArFioA8AKnCwAPGFgB2AE3SoYUYGIDmAPjgBeOAG86cOADdkAGwCuWAFxwBAMjgy5i+johgYwCGNTmtOnfuNmLDnAAvgDaALr0QfR0BEZi+K5icADCJJBi4jD8giLiUjay8soAFGBcHhzl-EoAlJracJQwRlBJYkYGBpEMAPQ9cACiUMRQ6EZgcDAQegCsAHQAzItwAEZG8GIQ8MgEolCzcwAsczN0PKngbpl6hiaqZAQQEGRwTi5uqKoaIRq+JuZkFbIKBkIJhIJwHpKejnNJXMTwP5Ye5AkGvZyJT7fX63fyA4Gg8GQ6FAA) with `T & string`, but does not work anymore in TS 5.4+.

File: packages/components/src/select-control/index.tsx
Comment (mikeybinns) [reply packages/components/src/select-control/index.tsx]: Technically, this isn't quite true because the select could be edited by someone using dev tools, however, I think this is fine because
1. If you edit a website with dev tools, it's your fault you broke the code, not the code maintainer
2. The likelihood of this happening is next to none

Comment (mikeybinns) [reply packages/components/src/select-control/index.tsx]: Would you be willing to use `OptionType` instead of `T` here? `T` doesn't really mean anything, and while it's a common practice in TypeScript, I see it as equivilent to naming a variable like `const a = 'something'`.

**Review by mikeybinns — APPROVED** [review]
I have some personal nit-picks, but ultimately approve it because I'd be fine if those were ignored :)

**Comment by mikeybinns** [comment]
One thing that's missing from my PR is the automated tests, perhaps it's worth copying those over?

File: packages/components/src/date-time/time/index.tsx
Comment (mikeybinns) [reply packages/components/src/date-time/time/index.tsx:1]: I agree with what you're saying. It would likely be a good thing to document how to specify a loose type (by using `SelectControl<string>()`) so that users who don't want to tighten their other code (in this case, the `format` function) can still have the same level of leniency as they previously had, and then this change empowers users to use tighter types if they like.

Though, saying that, I'm not sure where this would be documented. In the `SelectControl` docs as a note perhaps?

File: packages/components/src/select-control/types.ts
Comment (ciampo) [inline packages/components/src/select-control/types.ts]: Should we also use the `NoInfer` utility for the `onChange`'s arguments, in case a consumer of the component explicitly types `onChange`?

File: packages/components/src/date-time/time/index.tsx
Comment (ciampo) [reply packages/components/src/date-time/time/index.tsx:1]: > I just want to be sure that this type tightening is a net positive and not a nuisance 😅

This is definitely a big aspect to keep in mind. Overall, I think this change is a net positive — I hope that in the future we'll have a way to always apply the strictness, and not only when the `options` array is immutable.

We should probably write a dev note for this change, to help consumers of the package deal with this change correctly (especially the ones who need to support multiple versions of Wordpress / Gutenberg).

Comment (mirka) [reply packages/components/src/date-time/time/index.tsx:1]: I added a Dev Note so we at least have something to link to if anybody asks. If we do get some questions, we might need to document it somewhere more permanent. But in general I don't think we should have "TypeScript documentation" aside from our actual type files.

File: packages/components/src/select-control/index.tsx
Comment (mirka) [reply packages/components/src/select-control/index.tsx]: I changed them to `V` (for `value` — it's not actually the option type), does that read a bit better? We don't have official naming conventions for this project yet, so eventually we may decide on a prefixed convention like `TValue`. But to write out a full name like `Value`, I think it's harder to signal that it's a generic. And `ValueType` is redundant because everything is a type 😅 We can revisit, but I hope `V` is a good compromise for now.

File: packages/components/src/select-control/types.ts
Comment (mirka) [reply packages/components/src/select-control/types.ts]: Good one, I added the constraints (and type tests) for those too.

**Comment by mirka** [comment]
> One thing that's missing from my PR is the automated tests, perhaps it's worth copying those over?

You're right, I brought over your tests and reformatted them a bit to mesh with our [existing type tests](https://github.com/search?q=repo%3AWordPress%2Fgutenberg+%22static+typ%22+path%3A%22test%22&type=code) (ad20004). I tried to pare it down so we have a good cost/benefit balance between safety and maintenance burden — hope that's reasonable.

**Comment by mirka** [comment]
I discussed this with a few other folks and nobody is outright opposed, so I think we can move forward to an official review so this can be merged 🙏

File: packages/components/src/select-control/types.ts
Comment (jsnajdr) [reply packages/components/src/select-control/types.ts]: Can it be demonstrated when the `NoInfer` can be useful? A new test case that fails without it? I don't understand very well what it does.

Comment (mikeybinns) [reply packages/components/src/select-control/types.ts]: Sure, Without `NoInfer` Typescript will infer the generic type from the value prop as well as the options prop.

What this means is it won't detect that your value is not one of the provided options, and therefore won't provide an error for it.

I've created a [simple TS Playground example](https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABFApgZygdRlAFnEKAOTgEkxgUAnAHgBVEUAPVMAEzUQypjAHMAfAAoAbgEMANiBQAuRHQA0icVPRy6AbQC6AShkMA3oioooIKkhXTEAXwBQEBBkQEoAB0KIAvMnRYc+IQk5JRUQgBE3Lx84Uoa4SowbLGI4XBusE7hugDcdgD0+YjFiAB6APx2dqCQmUioGNh4wRTU9IwsKOycUfzCVrKILaH0AkoDaOraeobGpuaWktb2jmDOrh5QAEzevo0Bw9QRvTFxCZJJKWkZ8GvZOnmFJWWVQA) to illustrate the difference for you. Hope that helps 🙂

Comment (jsnajdr) [reply packages/components/src/select-control/types.ts]: Thanks, this is nice 🙂 I didn't realize that passing something as `value` can expand the `T` type to also include the type of `value`. But we want the type to be checked against `options` instead.

Comment (tyxla) [reply packages/components/src/select-control/types.ts]: Nice to see a clever usage of `NoInfer` 👍 I had read about it, but struggled to think of a good use case for it.

File: packages/components/src/select-control/index.tsx
Comment (DaniGuardiola) [inline packages/components/src/select-control/index.tsx]: I see what you were going for here, but I think I'd simplify the return type to just `React.JSX.Element | null`. That's the only thing that a component can ever return anyway, and it looks clearer to me this way. I don't really mind though, this is a nit.

**Review by DaniGuardiola — APPROVED** [review]
Good stuff! Straightforward implementation. Nice touch with the type tests, though it makes me wonder if we should add standards for type testing and/or support for tooling like https://github.com/MichiganTypeScript/type-testing - the eslint ignore comments for jest feel a bit like a hack.

As a reference, here's an example from a personal project where I do extensive type tests: https://github.com/DaniGuardiola/rpc-anywhere/blob/main/src/tests/types-test.ts

Other than that, I have no notes (other than a minor nit)!

File: packages/components/src/select-control/index.tsx
Comment (mirka) [reply packages/components/src/select-control/index.tsx]: That makes sense, thanks 🙏 [2f195c2](https://github.com/WordPress/gutenberg/pull/64069/commits/2f195c27ecf036e1f7359323fbaf87e53478071c)

**Comment by mirka** [comment]
> Nice touch with the type tests, though it makes me wonder if we should add standards for type testing and/or support for tooling like https://github.com/MichiganTypeScript/type-testing - the eslint ignore comments for jest feel a bit like a hack.

@DaniGuardiola It has come up, most recently in https://github.com/WordPress/gutenberg/pull/62400#issuecomment-2202602135. I don't think we're quite at the point where we need dedicated type test tooling, as the downsides are still outweighing the upsides in my opinion (e.g. having to manage the tooling package updates, forcing contributors to learn yet another testing library, etc).

But I do think we should start to document some standards/conventions like you suggest, as this is the fourth time we're [doing](https://github.com/search?q=repo%3AWordPress%2Fgutenberg+%22static+typ%22+path%3A%22%2Ftest%2F%22&type=code) static type tests here!

**Comment by mikeybinns** [comment]
Thanks for all your work on this @mirka ☺️

**Comment by mirka** [comment]
@mikeybinns Thanks for all your work as well 🙌

File: packages/components/src/select-control/types.ts
Comment (meteorlxy) [reply packages/components/src/select-control/types.ts]: Also found a good article about NoInfer: https://www.totaltypescript.com/noinfer

**Pull request merged**
